src/field_theory/finite/basic.lean

```diff
@@ -0,0 +1,205 @@
+    exact ⟨1⟩ },
+  cases is_cyclic.exists_generator (units K) with a ha,
+  calc univ.sum (λ (x : units K), (x^i : K)) = (range (order_of a)).sum (λ k, ((a^k)^i : K)) :
+  begin
+    symmetry,
+    refine sum_bij (λ i hi, a^i) (λ _ _, mem_univ _) (λ _ _, by rw units.coe_pow) _ _,
+    { intros i j hi hj h, rw [mem_range] at hi hj,
+      exact pow_injective_of_lt_order_of a hi hj h, },
+    { intros x hx, specialize ha x,
+      rwa [mem_gpowers_iff_mem_range_order_of, mem_image] at ha,
+      rcases ha with ⟨i, hi, rfl⟩, exact ⟨i, hi, rfl⟩ }
+  end
+    ... = geom_series (a^i : K) (q-1) :
+  begin
+    rw [order_of_eq_card_of_forall_mem_gpowers ha, card_units],
+    apply sum_congr rfl, intros k hk, rw [← pow_mul, mul_comm, pow_mul]
+  end
+    ... = if (q - 1) ∣ i then q - 1 else 0 :
+  begin
+    split_ifs with H H,
+    { rcases H with ⟨d, rfl⟩,
+      have aux : (λ (i:ℕ), ((a : K) ^ ((q - 1) * d)) ^ i) = λ i, 1,
+      { funext i, rw [pow_mul, pow_card_sub_one_eq_one _ (units.ne_zero _), one_pow, one_pow], },
+      rw [geom_series_def, aux, sum_const, card_range, add_monoid.smul_one,
+        nat.cast_sub, nat.cast_one],
+      exact le_trans hq (nat.pred_le _), },
+    { have key := geom_sum_mul (a^i : K) (q-1),
+      have hai : (a^i : K) ≠ 0, { rw ← units.coe_pow, apply units.ne_zero },
+      rw [pow_card_sub_one_eq_one _ hai, sub_self] at key,
+      replace key := eq_zero_or_eq_zero_of_mul_eq_zero key,
+      rw classical.or_iff_not_imp_right at key, apply key, contrapose! H,
+      rw [← card_units, ← order_of_eq_card_of_forall_mem_gpowers ha],
+      apply order_of_dvd_of_pow_eq_one,
+      rwa [units.ext_iff, units.coe_pow, units.coe_one, ← sub_eq_zero], }
+  end
+end
+
+theorem card (p : ℕ) [char_p K p] : ∃ (n : ℕ+), nat.prime p ∧ q = p^(n : ℕ) :=
+have hp : nat.prime p, from char_p.char_is_prime K p,
+have V : vector_space (zmodp p hp) K, from {..zmod.to_module'},
+let ⟨n, h⟩ := @vector_space.card_fintype _ _ _ _ V _ _ in
+have hn : n > 0, from or.resolve_left (nat.eq_zero_or_pos n)
+  (assume h0 : n = 0,
+  have q = 1, by rw[←nat.pow_zero (fintype.card _), ←h0]; exact h,
+  have (1 : K) = 0, from (fintype.card_le_one_iff.mp (le_of_eq this)) 1 0,
+  absurd this one_ne_zero),
+⟨⟨n, hn⟩, hp, fintype.card_fin p ▸ h⟩
+
+theorem card' : ∃ (p : ℕ) (n : ℕ+), nat.prime p ∧ q = p^(n : ℕ) :=
+let ⟨p, hc⟩ := char_p.exists K in ⟨p, @finite_field.card K _ _ p hc⟩
+
+@[simp] lemma cast_card_eq_zero : (q : K) = 0 :=
+begin
+  rcases char_p.exists K with ⟨p, _char_p⟩, resetI,
+  rcases card K p with ⟨n, hp, hn⟩,
+  simp only [char_p.cast_eq_zero_iff K p, hn],
+  conv { congr, rw [← nat.pow_one p] },
+  exact nat.pow_dvd_pow _ n.2,
+end
+
+/-- The sum of x^i as x ranges over a finite field of cardinality q is equal to 0 if i < q-1. -/
+lemma sum_pow_lt_card_sub_one (i : ℕ) (h : i < q - 1) :
+  univ.sum (λ x, x^i) = (0:K) :=
+begin
+  have hq : 0 < q - 1,
+  { rw [← card_units, fintype.card_pos_iff],
+    exact ⟨1⟩ },
```

Comment:
This isn't needed after other suggested changes.

Reply:
Still a TODO